### ISSUE TITLE
Code stuck on "initalizing ddp" when using more than one gpu 

### ISSUE THREAD

**Issue by JosephGatto** [PyTorchLightning/pytorch-lightning#4612]
## 🐛 Bug

I am trying to run a pytorch lightning model on a 4-GPU node. In my trainer, if I specify 

```
pl.Trainer(gpus=[0])
```
It runs fine. However, once I add another GPU
```
pl.Trainer(gpus=[0,1,2,3])
```
I get this output:

GPU available: True, used: True
TPU available: False, using: 0 TPU cores
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/4
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/4
initializing ddp: GLOBAL_RANK: 2, MEMBER: 3/4
initializing ddp: GLOBAL_RANK: 3, MEMBER: 4/4

And the model just hangs there forever. I have tried this with only 2 GPUs and get the same behavior. 

Any idea why this may happen? I have tried with both ddp and ddp_spawn.

 - PyTorch Version-- tried both 1.4 and 1.7
 - OS-- Linux
 - Installed with pip
 - Python version: 3.8.5
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: NVIDIA K80s

**Comment by edenlightning** [comment]
Hey! Can you try to reproduce using [our simple boring model](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py)?

Just to verify the bug isn't something in your model.

**Comment by JosephGatto** [comment]
Thanks for the response @edenlightning. This morning (about 12 hrs after my last attempt), I ran the simple boring model and it ran. I also added gpu=[0] and it worked. When I added gpu=[0,1] it worked the first time I ran it. I successfully ctrl+c'd out of it, then tried to run it again and it hangs forever. nvidia-smi shows no processes running and I can continually run the single gpu approach after this. However, anytime I do gpu > 1 from now on it does not run. 

Thus, it seems like something bad happens when I ctrl+c a ddp process that blocks it from happening again. Any idea what that might be?

**Comment by JosephGatto** [comment]
Update:  1) Have had my system admin restart this GPU node and it didn't run again. Not sure how it got through that one time.  2) None of the accelerators ['ddp', 'ddp_spawn', 'dp'] run for me when using gpu=[0,1] on boring model. 

Just curious, could this be a related issue? https://github.com/pytorch/pytorch/issues/1637 I am working on a 4-K80 node.

**Comment by JosephGatto** [comment]
@edenlightning any ideas about what the problem might be? ddp still isnt working on the boring model and ddp_spawn gives the pickle error on the boring model ... i am very stuck.

**Comment by edenlightning** [comment]
@justusschock mind taking a look?

**Comment by justusschock** [comment]
Hi @JosephGatto I am sorry that I cannot reproduce this, since I don't have these kind of GPUs. But I can try to guide you during troubleshooting.

Yes https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158 seems to be related. Have you tried the steps mentioned there to track down the problem?

**Comment by JosephGatto** [comment]
Hi @justusschock thanks for offering your help! Sadly, this did not work. Any other ideas?

**Comment by SohamTamba** [comment]
Hi, 
I had this problem to while running PL on my university's SLURM cluster. I was trying to use DDP on 4 GPUs.

Following the [tutorial](https://pytorch-lightning.readthedocs.io/en/latest/slurm.html), I set tasks per node to 4 (number of GPUs) and it hung on Initializing DDP.

I solved this by setting tasks per node to 8.

**Comment by justusschock** [comment]
@JosephGatto When you ctrl+c the dip process, do they still appear in `Nvidia-smi` as zombie processes?

**Comment by JosephGatto** [comment]
@justusschock Correct. I can't even ctrl+c I have to usually ctrl+z and then manually kill the process when I use ddp.

**Comment by justusschock** [comment]
What happens I you try to ctrl+c?

**Comment by JosephGatto** [comment]
@justusschock nothing, it just stays frozen. I am forced to ctrl+z.

**Comment by justusschock** [comment]
Even ctrl+c multiple times does not work?

**Comment by JosephGatto** [comment]
@justusschock correct

**Comment by yikuanli** [comment]
I have similar problem, and I m using the university's cluster, the exact same problem, hope someone can help out.

**Comment by Stellakats** [comment]
I have the exact same problem. Any idea of what could fix this?

**Comment by alexionby** [comment]
I faced the same problem. Looks like it somehow connected to zeromq (?)
```
Traceback (most recent call last):
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/traitlets/config/application.py", line 663, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-124>", line 2, in initialize
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/ipykernel/kernelapp.py", line 567, in initialize
    self.init_sockets()
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/ipykernel/kernelapp.py", line 271, in init_sockets
    self.shell_port = self._bind_socket(self.shell_socket, self.shell_port)
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/ipykernel/kernelapp.py", line 218, in _bind_socket
    return self._try_bind_socket(s, port)
  File "/home/alex/anaconda3/envs/pt/lib/python3.8/site-packages/ipykernel/kernelapp.py", line 194, in _try_bind_socket
    s.bind("tcp://%s:%i" % (self.ip, port))
  File "zmq/backend/cython/socket.pyx", line 550, in zmq.backend.cython.socket.Socket.bind
  File "zmq/backend/cython/checkrc.pxd", line 26, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Address already in use
```

**Comment by JosephGatto** [comment]
@alexionby I am no expert but an 'Address already in use' error in my pytorch-lightning experience has been related to an occupied port or ip address.

**Comment by alexionby** [comment]
@JosephGatto, I figured out that in my case it was connected to jupyter-notebook and ZMQ. As python script it works nice. My case closed.

**Comment by JosephGatto** [comment]
Hi @edenlightning @justusschock,

Does PyTorch-lightning support compute capability 3.7? One of the HPC specialists who manage my compute cluster tried debugging this today and said the issue was isolated to the K80 nodes and that he got it to work on other nodes that used compute capability 7.0. 

Note: The K80s failed even after a driver update and he said that all GPUs passed to PyTorch-lightning were working very hard, but whatever process pushes the workload to the GPU is not returning to the host.  

Thanks again.

**Comment by justusschock** [comment]
@JosephGatto We support whatever pytorch supports. we don't do anything related to specific cuda compute capabilities.
But even if the prebuilt binaries of pytorch don't support your GPU, you should be able to compile it yourself

**Comment by MarsSu0618** [comment]
@JosephGatto 
Did you solve it? Because i have same issues. It stucks when i set ddp ...

**Comment by shreyaskamathkm** [comment]
I have the same problem. Weirdly, it works well with V100 and P100 GPUs. But when I try using Tesla T4 GPUs, the code hangs.

**Comment by yorickvanzweeden** [comment]
I fixed it by adding `auto_select_gpus=True` as trainer_flag. I initially selected GPUs [0,1] which were in use by others.

**Comment by pgmikhael** [comment]
Also found it to work on V100 but encountered similar problem with A100 GPUs.

**Comment by jwohlwend** [comment]
On A100 GPU with cuda 11.0 and pytorch 1.7, vanilla pytorch DDP (without lightning) works fine, but lightning hangs on "initializing ddp: GLOBAL_RANK: 0". Interestingly in our setup lightning works with GPUs [0, 1] and [2, 3] but not with [1, 2].  Vanilla pytorch DDP works in all cases. Any idea what's going on?

**Comment by justusschock** [comment]
@jwohlwend Thanks for digging in!

**Comment by jwohlwend** [comment]
@justusschock No prob. This being said, it would be good if Lightning could allow overriding nccl with gloo. This was also talked about here #5355

**Comment by awaelchli** [comment]
Yes, I can confirm also it's a pytorch issue, we have seen it before https://github.com/PyTorchLightning/pytorch-lightning/issues/5264#issuecomment-751323437
The solution is to just avoid pytorch 1.7 and go with 1.6 or 1.8 :)

**Comment by jwohlwend** [comment]
@awaelchli Interesting, I did try 1.8 and it didn't work :/ And 1.6 i don't think supports A100 GPUs, does it?

**Comment by JosephGatto** [comment]
So for me this was a GPU problem. I don't manage my compute cluster but the system admin described the fix as a GPU configuration issue. This was for the K80s. Sorry I don't have more details!

**Comment by awaelchli** [comment]
Oh, maybe you needed the iommu="soft" setting. See this pytorch issue from 2017, https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158, I had the same problem back then. 
or maybe `NCCL_P2P_DISABLE=1`
maybe worth a try also for @jwohlwend

**Comment by shimengfeng** [comment]
Is there a solid solution yet? I am using p3.8xlarge and encountering the same issue as stuck at initializing ddp step.

**Comment by awaelchli** [comment]
@shimengfeng unfortunately no. I'm not able to reproduce based on all reported conditions. What worked for me is skipping pytorch 1.7

**Comment by shimengfeng** [comment]
I am currently using torch 1.6 and lightning 1.1.5, and I still get stuck at the initialization step. I can successfully run accelerator = 'dp' but just not 'ddp'. Am I suppose to change anything in my code to make it work? I am under the impression that I can directly change the accelerator and it should run. Also, I am running the code on databricks platform, with their p3.8xlarge. Not sure whether this will matter?

**Comment by shashank2000** [comment]
@shimengfeng exact same config and exact same issue - any luck?

**Comment by pgmikhael** [comment]
Potentially helpful, but certainly not ideal, forcing a switch to using gloo helped in my case -- setting `torch_backend` to gloo and switching off the if statement here [https://github.com/PyTorchLightning/pytorch-lightning/blob/1.1.6/pytorch_lightning/plugins/ddp_plugin.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/1.1.6/pytorch_lightning/plugins/ddp_plugin.py)

**Comment by shimengfeng** [comment]
> @shimengfeng exact same config and exact same issue - any luck?

@shashank2000 I was told that databricks does not support ddp as of now, and I ended up using horovod and it worked for me that way.

**Comment by karthi0804** [comment]
I have the same problem. Multi-node training works well with Pytorch 1.7 DDP (using both the backend gloo and nccl). But with Pytorch lightning 1.1.4, it's getting stuck. Not sure if something wrong with my slurm setting. Please find few details of 2 nodes training (one has 2 gpus and other one has 1 gpu). Though the first node has two gpus, I m using one to experiment DDP. Someone pls help.
The pl trainer params:
```
trainer = pl.Trainer(num_nodes=2, gpus=1, accelerator='ddp')
```
slurm bash file:
```
#!/bin/bash -l

# SLURM SUBMIT SCRIPT
#SBATCH --nodes=2
#SBATCH --gres=gpu:1
#SBATCH --ntasks-per-node=1
#SBATCH --mem=0
#SBATCH --time=0-02:00:00

# -------------------------
# OPTIONAL
# -------------------------
# debugging flags (optional)
export NCCL_DEBUG=INFO
export PYTHONFAULTHANDLER=1

# PyTorch comes with prebuilt NCCL support... but if you have issues with it
# you might need to load the latest version from your  modules
# module load NCCL/2.4.7-1-cuda.10.0

# on your cluster you might need these:
# set the network interface
export NCCL_SOCKET_IFNAME=enp0s31f6
# -------------------------

# random port between 12k and 20k
export MASTER_PORT=$((12000 + RANDOM % 20000))

# run script from above
python main.py
```

**Comment by PengyuWang** [comment]
I met the same problem as well. 

1, DDP and DP works well in PyTorch.
2, DP works well in PyTorch Lightning, but sometimes DDP stuck on "initalizing ddp".

Please fix it.

**Comment by SapirWeissbuch** [comment]
I'm having the same problem in the university's slurm. I'm using PyTorch 1.6 (as mentioned above) but still gets stuck on `initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2`

**Comment by SapirWeissbuch** [comment]
> I'm having the same problem in the university's slurm. I'm using PyTorch 1.6 (as mentioned above) but still gets stuck on `initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2`

Update: switching from `srun` to `sbatch` in Slurm fixed the issue.

**Comment by Akanni96** [comment]
I've also faced a problem with code stuck on "initializing ddp". After much work, I solved my problem by simply adding `num_sanity_val_steps=0` as an argument to `Trainer(...)`

**Comment by calebclayreagor** [comment]
I had this same issue, and as per @pgmikhael, switching the torch backend to 'gloo' solves the issue

**Comment by andrewssobral** [comment]
I had the same issue in my K80 GPU Cluster, and I solved it by using the Gloo interface instead of NCCL as you can see in this discussion here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/6509 (with source code)

**Comment by calebclayreagor** [comment]
As a newbie, I did not know to check my GPU processes using `nvidia-smi` but it turns out my problem was simply ghost processes that needed to be killed (`kill -9 <pid>`) before initializing ddp again. Closed for me.

**Comment by awaelchli** [comment]
@calebclayreagor that also happens to me from time to time. We have recently identified some problems where processes don't shut down properly. The changes that this happens will be greatly reduced after #6864 is merged. 

You can kill all stuck python processes with `pkill -9 python`

**Comment by hoangtnm** [comment]
> Yes, I can confirm also it's a pytorch issue, we have seen it before [#5264 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/issues/5264#issuecomment-751323437)
> The solution is to just avoid pytorch 1.7 and go with 1.6 or 1.8 :)

@awaelchli I have tried pytorch 1.6 on a computer with 2x 1080Ti, and it seems work fine, but I still witness this issue with pytorch 1.8.1 or 1.7.

**Comment by haossr** [comment]
Any updates on this?

**Comment by hoangtnm** [comment]
> Any updates on this?

This issue still exists with the latest versions of Pytorch and PyTorch Lightning.

**Comment by awaelchli** [comment]
for everyone finding this issue or still having problems, make sure you run with `NCCL_DEBUG=INFO` and include it in any bug reports (here or anywhere else). DDP/NCCL hanging can be caused by many things. Setting the `NCCL_DEBUG=INFO` environment variable often tells us why.

**Comment by twsl** [comment]
Same problem here, using Databricks with `Standard_NC12` node. Using `ddp_spawn` it crashes after showing  `LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]`. Tested with the simple boring model:
<details>
<summary>Exception</summary>

```
---------------------------------------------------------------------------
ProcessExitedException                    Traceback (most recent call last)
<command-1357939202606131> in <module>
----> 1 trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)

/databricks/python/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloaders, val_dataloaders, datamodule, train_dataloader)
    551         self.checkpoint_connector.resume_start()
    552 
--> 553         self._run(model)
    554 
    555         assert self.state.stopped

/databricks/python/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py in _run(self, model)
    916 
    917         # dispatch `start_training` or `start_evaluating` or `start_predicting`
--> 918         self._dispatch()
    919 
    920         # plugin will finalized fitting (e.g. ddp_spawn will load trained model)

/databricks/python/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py in _dispatch(self)
    984             self.accelerator.start_predicting(self)
    985         else:
--> 986             self.accelerator.start_training(self)
    987 
    988     def run_stage(self):

/databricks/python/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py in start_training(self, trainer)
     90 
     91     def start_training(self, trainer: "pl.Trainer") -> None:
---> 92         self.training_type_plugin.start_training(trainer)
     93 
     94     def start_evaluating(self, trainer: "pl.Trainer") -> None:

/databricks/python/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py in start_training(self, trainer)
    156 
    157     def start_training(self, trainer):
--> 158         mp.spawn(self.new_process, **self.mp_spawn_kwargs)
    159         # reset optimizers, since main process is never used for training and thus does not have a valid optim state
    160         trainer.optimizers = []

/databricks/python/lib/python3.8/site-packages/torch/multiprocessing/spawn.py in spawn(fn, args, nprocs, join, daemon, start_method)
    228                ' torch.multiprocessing.start_processes(...)' % start_method)
    229         warnings.warn(msg)
--> 230     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')

/databricks/python/lib/python3.8/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    186 
    187     # Loop on join until it returns True or raises an exception.
--> 188     while not context.join():
    189         pass
    190 

/databricks/python/lib/python3.8/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    137                 )
    138             else:
--> 139                 raise ProcessExitedException(
    140                     "process %d terminated with exit code %d" %
    141                     (error_index, exitcode),

ProcessExitedException: process 0 terminated with exit code 1
```

</details>

PyTorch Version: 1.9.0+cu102 (Lightning 1.4.1)
OS: Linux
Python version: 3.8.8
CUDA/cuDNN version: 11
GPU models and configuration: 2xNVIDIA Tesla K80

**Comment by sordonia** [comment]
same for me, but the code doesn't get stuck with commit 03bb389b218084f365b922f2a50133a5aaaadaf0

**Comment by justusschock** [comment]
cc @ethanwharris who introduced this. Do you remember why you removed it?

**Comment by ethanwharris** [comment]
@justusschock I think @sordonia is saying that the break happened _since_ that commit rather than in it. At least, I don't recognize that change and can't find it in the diff from that PR. It looks like the line change that @sordonia is referring to happened in #7480 cc @awaelchli

**Comment by sordonia** [comment]
Yeah, I don't know *when* the break happened, I just diff-ed the commit I know to work with current master which doesn't work for me.

**Comment by awaelchli** [comment]
Hey, thanks for reporting it. I believe the conclusion of your observation is not correct:

> removing os.environ... makes all created subprocesses go into _call_children_scripts() and then fail with assert local_rank == 0 causing the block, a solution would be reinstating the os check or adding:

The decision here depends solely on the cluster environment. If the cluster has spawned the processes, Lightning will not spawn them. If cluster has NOT spawned them, Lightning will do so if requested. 

In the latter case, local_rank=0 holds true since at this point the only process present must be the main one! 

Let's move this discussion to a separate issue and then we will request more information of how you are launching the ddp job. Thanks for your understanding.

**Comment by aleSuglia** [comment]
I'm experiencing the same issue. The trainer seems to be working with DP though...

**Comment by saitjinwon** [comment]
I am experienceing the same problem. 

Except that it does not work in 'dp' setting as well as in 'ddp' setting'

I ran the cifar-10 example with multiple gpus


``
import os

import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
from pl_bolts.datamodules import CIFAR10DataModule
from pl_bolts.transforms.dataset_normalizations import cifar10_normalization
from pytorch_lightning import LightningModule, seed_everything, Trainer
from pytorch_lightning.callbacks import LearningRateMonitor
from pytorch_lightning.loggers import TensorBoardLogger
from torch.optim.lr_scheduler import OneCycleLR
from torch.optim.swa_utils import AveragedModel, update_bn
from torchmetrics.functional import accuracy

seed_everything(7)

PATH_DATASETS = os.environ.get('PATH_DATASETS', '.')
AVAIL_GPUS = torch.cuda.device_count()
BATCH_SIZE = 256 if AVAIL_GPUS else 64
NUM_WORKERS = int(os.cpu_count() / 2)

train_transforms = torchvision.transforms.Compose([
    torchvision.transforms.RandomCrop(32, padding=4),
    torchvision.transforms.RandomHorizontalFlip(),
    torchvision.transforms.ToTensor(),
    cifar10_normalization(),
])

test_transforms = torchvision.transforms.Compose([
    torchvision.transforms.ToTensor(),
    cifar10_normalization(),
])

cifar10_dm = CIFAR10DataModule(
    data_dir=PATH_DATASETS,
    batch_size=BATCH_SIZE,
    num_workers=NUM_WORKERS,
    train_transforms=train_transforms,
    test_transforms=test_transforms,
    val_transforms=test_transforms,
)

def create_model():
    model = torchvision.models.resnet18(pretrained=False, num_classes=10)
    model.conv1 = nn.Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    model.maxpool = nn.Identity()
    return model

class LitResnet(LightningModule):

    def __init__(self, lr=0.05):
        super().__init__()

        self.save_hyperparameters()
        self.model = create_model()

    def forward(self, x):
        out = self.model(x)
        return F.log_softmax(out, dim=1)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        self.log('train_loss', loss)
        return loss

    def evaluate(self, batch, stage=None):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        preds = torch.argmax(logits, dim=1)
        acc = accuracy(preds, y)

        if stage:
            self.log(f'{stage}_loss', loss, prog_bar=True)
            self.log(f'{stage}_acc', acc, prog_bar=True)

    def validation_step(self, batch, batch_idx):
        self.evaluate(batch, 'val')

    def test_step(self, batch, batch_idx):
        self.evaluate(batch, 'test')

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(
            self.parameters(),
            lr=self.hparams.lr,
            momentum=0.9,
            weight_decay=5e-4,
        )
        steps_per_epoch = 45000 // BATCH_SIZE
        scheduler_dict = {
            'scheduler': OneCycleLR(
                optimizer,
                0.1,
                epochs=self.trainer.max_epochs,
                steps_per_epoch=steps_per_epoch,
            ),
            'interval': 'step',
        }
        return {'optimizer': optimizer, 'lr_scheduler': scheduler_dict}

model = LitResnet(lr=0.05)
model.datamodule = cifar10_dm

trainer = Trainer(
    progress_bar_refresh_rate=10,
    max_epochs=30,
    gpus=AVAIL_GPUS,
   accelerator='dp',
    logger=TensorBoardLogger('lightning_logs/', name='resnet'),
    callbacks=[LearningRateMonitor(logging_interval='step')],
)

trainer.fit(model, cifar10_dm)
trainer.test(model, datamodule=cifar10_dm)

``

The output is hanged after working for just one step of training_step(one batch for each gpu).

Also, even if I press Ctrl+C multiple times, it does not halt. So I had to kill the process by looking up in htop.

**Comment by tchaton** [comment]
Hey @saitjinwon,

I have tried your script using PyTorch Lightning master on using both dp and ddp on 2 gpus and it seems to work fine.

Would you mind trying out master ?

```py
pip install git+https://github.com/PyTorchLightning/pytorch-lightning.git
```

Best,
T.C

**Comment by H-B-L** [comment]
I have the same issue. I filed the issue here `#10471`.

**Comment by carmocca** [comment]
Closing. If you are a future reader and none of the existing discussions helped you, please open a new issue with details and reproduction for your hang.

**Comment by raymondconfidence** [comment]
> Hi, I had this problem to while running PL on my university's SLURM cluster. I was trying to use DDP on 4 GPUs.
> 
> Following the [tutorial](https://pytorch-lightning.readthedocs.io/en/latest/slurm.html), I set tasks per node to 4 (number of GPUs) and it hung on Initializing DDP.
> 
> I solved this by setting tasks per node to 8.

Worked for me.
Thanks for sharing

**Comment by jshilong** [comment]
Hi guys!
I've encountered a similar issue recently.  My program hangs at 'initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2' when running a repository requiring an older PyTorch. I'm testing it with both the 4090 and A100 GPUs, and due to changes in the computing architecture, these two cards now have minimum CUDAToolkit version requirements. However, the older PyTorch installation in the repository doesn't provide any relevant error messages, such as 

'NVIDIA GeForce RTX 4090 with CUDA capability sm_89 is not compatible with the current PyTorch installation. The current PyTorch installation supports CUDA capabilities sm_37 sm_50 sm_60 sm_61 sm_70 sm_75 compute_37.' in the newer versions of PyTorch.

so you just need to update to a newer version of PyTorch with a compatible CUDAToolkit version. Hopefully, this can help some people out.

**Comment by ImNotPrepared** [comment]
Hi guys, I ran into the same problem, I am using slurm from university. just change --ntasks-per-node=1 solves my problem

**Comment by ahmadikalkhorani** [comment]
using srun before python  solved my problem: 

srun python train.py ...
